File: CodingSamples/DotNet/Advanced/AspNetCoreMvc/ControllerTest/DemoApp1/Tourism/Models/SiteModel.cs
using DemoApp.Tourism.Data;

namespace DemoApp.Tourism.Models;

public class SiteModel
{
    public IEnumerable<Visitor> GetVisitors()
    {
        using var site = new SiteDbContext();
        var selection = from t in site.Travellers
            where t.Id.Length > 3
            select new Visitor 
            {
                Name = t.Id,
                Stars = new string('*', t.Rating),
                Visits = t.Trips.Count,
                Recent = t.Trips.Max(x => x.Checkin)
            };
        return selection.ToList();
    }

    public void HandleVisit(string visitorName, int visitorRating)
    {
        using var site = new SiteDbContext();
        var traveller = site.Travellers.Find(visitorName);
        if(traveller is null)
        {
            traveller = new Traveller { Id = visitorName };
            site.Travellers.Add(traveller);
        }
        traveller.Trips.Add(new Trip());
        traveller.Rating = visitorRating;
        site.SaveChanges();
    }
}

File: CodingSamples/Java/Advanced/ServletAPI/WebServletTest/tours/src/main/java/app/tourism/models/SiteModel.java
package app.tourism.models;

import java.util.List;

import app.tourism.data.TravelerEntity;
import app.tourism.data.TripEntity;
import jakarta.persistence.EntityManagerFactory;
import jakarta.persistence.Persistence;

public class SiteModel {

    private static final EntityManagerFactory emf = Persistence.createEntityManagerFactory("app.data");

    public List<Visitor> getVisitors() {
        try(var em = emf.createEntityManager()){
            var query = em.createQuery("SELECT e FROM TravelerEntity e WHERE LENGTH(e.id) > 3", TravelerEntity.class);
            return query.getResultStream()
                .map(Visitor::fromTraveler)
                .toList();
        }
    }

    public boolean handleVisit(String visitorId, int visitorRating) {
        if(visitorId == null || visitorRating < 1 || visitorRating > 5)
            return false;
        try(var em = emf.createEntityManager()){
            var traveler = em.find(TravelerEntity.class, visitorId);
            if(traveler == null){
                traveler = new TravelerEntity();
                traveler.setId(visitorId);
            }
            traveler.setRating(visitorRating);
            var trip = new TripEntity();
            trip.setGuest(traveler);
            traveler.getTours().add(trip);
            var tx = em.getTransaction();
            tx.begin();
            em.persist(traveler);
            tx.commit();
            return true;
        }
    }
    
}
